### ISSUE TITLE
Code Explorer missed a form

### ISSUE THREAD

**Issue by AndrewM-** [rubberduck-vba/Rubberduck#827]
The form has an inconvenient name but it probably best to let you know so you can put it in a unit test or something.  I can live with this small issue.
![missedform](https://cloud.githubusercontent.com/assets/5060805/10416578/a60a64f2-705f-11e5-91d4-fc5a2cd14551.png)

**Comment by ckuhn203** [comment]
That's typically a parsing issue. Does the code in that form compile?

**Comment by retailcoder** [comment]
Definitely a parsing issue. Go to that form's code-behind, and right-click any procedure and select "Find all references" from the code pane's context menu.

This will initiate a parse, and if there's a parser error, a dockable toolwindow will pop up with what's usually a pretty cryptic ANTLR parser error message that's remotely related to where the parse error is - double-click that error message.

I expect it takes you to a line that does something like this:

    foo = SomeFunctionCallThatReturnsAnArray(bar, 123)(0)

This is a known parser issue that has now been fixed (thought not released yet) - there's a work-around:

    result = SomeFunctionCallThatReturnsAnArray(bar, 123)
    foo = result(0)

@ckuhn203 also makes a fair comment: the Rubberduck parser assumes VBA can compile the code. Beyond the few known parser issues, it's pretty much the only thing that could cause a given module to not be parsed.

**Comment by retailcoder** [comment]
Part of the problem is that there's a UX issue here: parser errors don't show up when it's the code explorer or code inspections calling into the parser, only when the `ParseProgress` window gets to show up.

**Comment by AndrewM-** [comment]
The code does compile.  I add some junk code to frmCollections, compile, delete the code and the compile again to ensure that the code is freshly compiled.  Can also force a fresh compile by starting the project from the command line with the /decompile switch.  

"Definitely a parsing issue. Go to that form's code-behind, and right-click any procedure and select 'Find all references' from the code pane's context menu". Using Code Explorer, none of my forms appears to have any references.  The parser error I get are shown below.  All of these errors are in code that I have brought in from other sources and I do not know how to modify the code to suppress the parser errors.  On inspection the code that doesn't parse looks fine.  

![parsererrors](https://cloud.githubusercontent.com/assets/5060805/10420146/ec06442a-70ce-11e5-9857-a0de7c9bb685.png)

Below are the offending lines with my comments.
Error 1:
#If Win64 Then 'for 64 bit office
    Declare PtrSafe Function abOpenClipboard Lib "user32" Alias "OpenClipboard" (ByVal hWnd As Long) As Long  (do compiler directives parse correctly?)

Error 2:
On Error GoTo Err_import:  (the colon is the issue, can fix this one)

Error 3:
    If f.buffer_pos >= &H4000 Then (no idea)

Error 4:
    If ((GetAttr(SourceDirectory) And vbDirectory) = vbDirectory) Then  (no idea)

Error 5:
    ' open file for reading with Create=False, Unicode=True (USC-2 Little Endian format)
    Set InFile = fso.opentextfile(tempFilePath, ForReading, False, TristateTrue)
There is something wrong with the above but I can see it yet.

**Comment by retailcoder** [comment]
"Find all references" was really just to force a parse from the code pane ;-)

I think most of the 'unexpected input' issues are in fact on the line immediately above the reported one.

The grammar handles `#If` conditional compilation, unless the logic is intertwined with `If` blocks or `For` loops. That's a *known issue* in the `VisualBasic6.g4` file we created `VBA.g4` from.

That original grammar file had errors; only recently we noticed it doesn't handle `#Const` tokens at all, and we've fixed quite an amount of other things, too. But although it's still buggy, it's far closer to correctly pick up VBA tokens than anything else we tried before, so we'll tweak it until we fix it - and this is why your input is so appreciated: we get to run our parser on hundreds of thousands of code lines (I like to think that ;-), and when we get feedback about parser errors it's an opportunity to take down another grammar bug!

Make a new VBA project (go to Excel and ctrl+n) and copy the offending code into an empty module (like the `Sheet1` code-behind), and try to reproduce the parsing issue with as little code as possible while keeping the code compilable; parser errors are often related to the line right above the reported one. Feel free to share more of the surrounding code if you need help narrowing down on a parser issue.

I wonder if error 4 has anything to do with the extraneous parentheses... but yeah I see nothing wrong there, I suspect the parser problem is higher up, as with the other errors.

**Comment by retailcoder** [comment]
Also... @ckuhn203 Access forms are `vbext_ct_document` components, can you confirm whether the VBE knows about a designer for them? We're only trying to get the designer from *form* components. *todo: find and link to the issue# this relates to*

**Comment by AndrewM-** [comment]
I will package up any code that does not parse and send it to you as a zipped accdb file.  I understand what it like trying to get user feedback on complex projects which is why I like to help.  My first instinct is to attach every issue with a database application, so I very seldom use Excel.  It would not be hard to parse all of my applications.  That would add another 50 K lines of experience.

**Comment by retailcoder** [comment]
Problem is, I don't have MS-Access :-)

**Comment by AndrewM-** [comment]
That may be solvable.  I use office 2010 because I don't like flat icons.  Office 2016 is also due out soon and I will have to get a copy of that or at least the ms-access part.  I use LibreOffice for everything else.  I have a legal copy of ms-access 2013 which I don' t use and I part with it, if it would help.  Could send it by dropbox.

**Comment by ckuhn203** [comment]
Yeah @retailcoder. The VBE knows about their designer.

**Comment by retailcoder** [comment]
I wonder then, if all we need to do is to also run the control-finding code on *document* modules with a name that starts with `Form_`...

**Comment by ckuhn203** [comment]
Possibly.

**Comment by retailcoder** [comment]
With precompiler directives handled, and the grammar/parser having no more known issues, I'm 99.99999% sure that code won't be a problem with v2.0 - closing the issue, but I'll reference here on a new one so that we remember to treat `Form_` document modules with an attached designer as, well, forms with controls.
